akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala

```diff
@@ -48,7 +53,13 @@ object ClusterShardingPersistenceSpec {
       entityId,
       emptyState = "",
       commandHandler = (state, cmd) ⇒ cmd match {
-        case Add(s) ⇒ Effect.persist(s)
+        case Add(s) ⇒
+          Effect.persist(s)
+
+        case cmd @ AddWithConfirmation(s) ⇒
+          Effect.persist(s)
+            .thenReply(cmd)(newState ⇒ Done)
```

Comment:
Compared to the current zero-api 
```
.thenRun(newState => 
  cmd.replyTo ! Done
)
```

I don't think this provides any value at all. You could just as easily forget to do the `thenReply`

Reply:
Well, it does provide the value of allowing for at least an improved discoverability on *where* to issue replies.

Reply:
That is true. On the other hand it requires some ceremony when defining the commands; have to implement a trait compared to defining the `replyTo` in the messages in whatever way you like. It also adds another way to send messages. I'm on the fence but would be ok with having this if others think it's needed.

Reply:
Unless there can be some kind of type checking that a reply is issued at all, I agree with @johanandren and @chbatey that this doesn't seem valuable.

I'm still clinging to a hope that there's some kind of (opt-in) pattern possible that could type check the reply (including the presence of a reply), at least for simple request-response flows that I think are the common case for persistent entities.

Reply:
But we can enforce the response, by requiring a specific type of effect to be returned: a ResponseEffect.

That is orthogonal though. `thenTell` is another option. I've seen cases where people respond *outside* of the effects, and that's not great.

Reply:
> requiring a specific type of effect to be returned: a ResponseEffect

Could you show me how to do that? That return type depends on the specific command type.

>  I've seen cases where people respond outside of the effects, and that's not great.

If command is not accepted you want to respond outside an effect, in my opinion. I see no point in doing that with `Effect.none.thenReply(cmd)(_ => Rejected)` when all it is `cmd.replyTo ! Rejected`.

There can be all kinds of variations, perhaps you want to validate the command by interacting with some other actors before replying. Perhaps you also want to sent invalid commands to the overlord for audit.

Reply:
>Could you show me how to do that? That return type depends on the specific command type.

```scala
sealed trait Cmd { type Response }

final case class AddPorsche(licensePlateNo: String) extends Cmd { type Response = Boolean }

final case class ShowPorsche(licensePlateNo: String) extends Cmd { type Response = String }

trait ResponseEffect

object Effects {
  def respondTo[T <: Cmd](command: T)(response: command.Response): ResponseEffect = ???
}

trait PersistentEntity {
  def onCommand(cmd: Cmd) : ResponseEffect
}

class PorscheCollector extends PersistentEntity {
  override def onCommand(cmd: Cmd): ResponseEffect =
    cmd match {
      case a @ AddPorsche(plate) => Effects.respondTo(a)(true)
      case s @ ShowPorsche(plate) => Effects.respondTo(s)("A really nice Porsche")
    }
}
```

Reply:
@patriknw `Effects.noResponse` and have those Cmd's which do not respond use NotUsed?

Reply:
That's a good suggestion. Enforcing it on all commands with explicit opt-out is possible.
I have implemented that and pushed to this PR, see PersistentBehaviorReplySpec: https://github.com/akka/akka/pull/25692/files#diff-deb775cb1ed6e5e768717073cc78711eR37